File: PHP/src/OneLang/One/Transforms/FillAttributesFromTrivia.php
<?php

namespace OneLang\One\Transforms\FillAttributesFromTrivia;

use OneLang\One\Ast\Types\SourceFile;
use OneLang\One\Ast\Types\IMethodBase;
use OneLang\One\Ast\Types\IHasAttributesAndTrivia;
use OneLang\One\Ast\Types\Package;
use OneLang\One\Ast\Types\IMethodBaseWithTrivia;
use OneLang\One\Ast\Statements\ForeachStatement;
use OneLang\One\Ast\Statements\ForStatement;
use OneLang\One\Ast\Statements\IfStatement;
use OneLang\One\Ast\Statements\Block;
use OneLang\One\Ast\Statements\WhileStatement;
use OneLang\One\Ast\Statements\DoStatement;
use OneLang\One\AstTransformer\AstTransformer;
use OneLang\One\Ast\Expressions\Expression;

class FillAttributesFromTrivia extends AstTransformer {
    function __construct() {
        parent::__construct("FillAttributesFromTrivia");
        
    }
    
    protected function visitAttributesAndTrivia($node) {
        $node->attributes = FillAttributesFromTrivia::processTrivia($node->leadingTrivia);
    }
    
    protected function visitExpression($expr) {
        return $expr;
    }
    
    static function processTrivia($trivia) {
        $result = Array();
        if ($trivia !== null && $trivia !== "") {
            $regex = new \OneCore\RegExp("(?:\\n|^)\\s*(?://|#|/\\*\\*?)\\s*@([A-Za-z0-9_.-]+) ?((?!\\n|\\*/|$).+)?");
            while (true) {
                $match = $regex->exec($trivia);
                if ($match === null)
                    break;
                if (array_key_exists($match[1], $result))
                    // @php $result[$match[1]] .= "\n" . $match[2];
                    // @python result[match[1]] += "\n" + match[2]
                    // @csharp result[match[1]] += "\n" + match[2];
                    // @java result.put(match[1], result.get(match[1]) + "\n" + match[2]);
                    $result[$match[1]] .= "\n" . $match[2];
                else
                    $result[$match[1]] = ($match[2] ?? "") === "" ? "true" : $match[2] ?? "";
            }
        }
        return $result;
    }
}

File: Python/OneLang/One/Transforms/FillAttributesFromTrivia.py
from onelang_core import *
import OneLang.One.Ast.Types as types
import OneLang.One.Ast.Statements as stats
import OneLang.One.AstTransformer as astTrans
import OneLang.One.Ast.Expressions as exprs

class FillAttributesFromTrivia(astTrans.AstTransformer):
    def __init__(self):
        super().__init__("FillAttributesFromTrivia")
    
    def visit_attributes_and_trivia(self, node):
        node.attributes = FillAttributesFromTrivia.process_trivia(node.leading_trivia)
    
    def visit_expression(self, expr):
        return expr
    
    @classmethod
    def process_trivia(cls, trivia):
        result = {}
        if trivia != None and trivia != "":
            regex = RegExp("(?:\\n|^)\\s*(?://|#|/\\*\\*?)\\s*@([A-Za-z0-9_.-]+) ?((?!\\n|\\*/|$).+)?")
            while True:
                match = regex.exec(trivia)
                if match == None:
                    break
                if match[1] in result:
                    # @php $result[$match[1]] .= "\n" . $match[2];
                    # @python result[match[1]] += "\n" + match[2]
                    # @csharp result[match[1]] += "\n" + match[2];
                    # @java result.put(match[1], result.get(match[1]) + "\n" + match[2]);
                    result[match[1]] += "\n" + match[2]
                else:
                    result[match[1]] = "true" if (match[2] or "") == "" else match[2] or ""
        return result